docs/sources/lids/_DRAFT-Index-Gateway-Sharding.md

```diff
@@ -0,0 +1,97 @@
+
+**Thread from [mailing list](https://groups.google.com/forum/#!forum/lokiproject):**
+
+---
+
+## Background
+
+This document tries to come up with a proposal on how to do a better sharding of data on the index gateways so we are able to scale the service horizontally to fulfill the increased need for metadata queries of big tenants.
+
+The index gateway service can be run in "simple mode", where an index gateway instance is responsible for handling, storing and returning requests for all indices for all tenants, or in "ring mode", where an instance is responsible for a subset of tenants instead of all tenants.
+
+On top of that, in order to achieve redundancy as well as spreading load, the index gateway ring uses by default a replication factor of 3.
+
+This means, before an index gateway client makes a request to the index gateway server, it first hashes the tenant ID and then requests a replication set for that hash from the index gateway ring. Due to the fixed replication factor (RF), the replication set contains three server addresses. On every request, a random server from that list is picked to then execute the request on.
+
+## Problem Statement
+
+The current strategy of sharding by tenant ID and having a replication factor fails in the long run, because even when running lots of index gateways, only a maximum of `n` instances could be utilized by a single tenant, where `n` is the value of the configured RF.
+
+Another problem is that the RF is fixed and the same for all tenants, independent of their actual size in terms of log volume or query rate.
+
+## Goals
+
+The goal of this document is to find a better sharding mechanism for the index gateway, so that there are no boundaries for scaling the service horizontally.
+
+* The sharding needs to account for the "size" of a tenant.
+* A single tenant needs to be able to utilize more than three index gateways.
+
+## Proposals
+
+### Proposal 0: Do nothing
+
+If we do not improve the sharding mechanism for the index gateways and leave it as it is, it will become more and more difficult to serve metadata queries for large tenant in a reasonable amount of time, proportionally to the demand for these queries.
+
+### Proposal 1: Dynamic replication factor
+
+Instead of using a fixed replication factor of 3, the RF can be derived from the amount of active members in the index gateway ring. That means that the RF would be a percentage of the available gateway instances. For example, a ring with 12 instances and 30% replication utilization would result in a RF of 3 (`floor(12*0.3)`). Scaling up to 18 instances would result in a RF of 5.
+
+This approach would solve the problem of horizontal scaling. However, it does not solve the problem of different tenant sizes. It also fails to ensure replication for availability when running a small number of instances, unless there is a fixed lower value for the RF. It also tends to over-shard data in large index gateway deployments.
```

Comment:
Does this also imply a resharding?

Reply:
Active re-sharding is not considered in any of the proposals. At the moment, index files are treated as transient/intermediate resources, because they can be re-downloaded for processing when the assignment changes.